working/0110-incremental-sound-nnbd/roadmap.md

```diff
@@ -0,0 +1,874 @@
+# Sound non-nullable (by default) types with incremental migration 
+
+Author: leafp@google.com
+
+This document proposes a roadmap for implementing a sound null tracking type
+system in Dart similar to what has been
+explored [previously](https://github.com/dart-lang/sdk/pull/28619/files).  The
+purpose of this document is not to describe the technical details of a proposal,
+but instead to argue for a specific set of high-level goals and design choices
+that define the key properties of the final system, and of the migration path to
+get there.  Specifically, we propose to aim for a system which allows for an
+incremental opt-in migration, and which is fully sound once all code in a
+compilation target has opted-in.
+
+## Motivation for non-nullability in Dart.
+
+This has been fairly well explored previously, so just in brief.  This is one of
+the most requested features in developer surveys, and one of the thing that
+developers most frequently mention missing from other languages.  Kotlin, Swift,
+C#, Typescript, and numerous other languages now have support for
+non-nullability.  Since Dart also has no primitive types or value types, it is
+possible that non-nullability could provide performance benefits as well.
+
+## Summary of proposed goals
+
+We propose to set the following goals for nullability tracking in Dart.
+
+- The type system should be incrementally sound.  That is, compilation should be
+possible with a mix of opted-in and non-opted-in code, but when all code in a
+compile has opted in, the compiler should be able to trust what the type system
+says about nullability.  When not all code has opted-in, it is possible that a
+non-nullably typed value will receive null.
+- During the migration, packages should be able to opt in without breaking
+  non-opted-in clients.
+- During the migration, packages should be able to opt in before packages that
+  they depend on have opted in.
+- During the migration, if a package opts in after all of its dependencies have
+already opted in (or if it correctly predicts and codes against the
+post-migration API of its dependencies) it only migrates once.
+
+## Proposal roadmap
+
+### Opting in
+
+Non-nullable types will be controlled by a language opt-in as deal with
+elsewhere.  In short, a library may opt-in via syntax in the code, or a package
+may opt-in in entirety.  Opting-in applies only to the library or package so
```

Comment:
I would be great if it was easy to detect when a whole package has opt'ed in. This would make the analysis on Pub much easier.

Reply:
I think should be fairly easy to do.